<head>
<!-- Generated by perlmod2www.pl -->
<title>
JobGraph documentation.
</title>
<link rel="stylesheet" type="text/css" href="../perl.css">
</head>
<body bgcolor="white">
<span id="modRootPath">MyBio::NGS</span>
<span id="modName">JobGraph</span>
<div id="sep"></div>
<a name="TOP" id="TOP"></a>
<table id="daArea" width="100%"><tr><td id="daEntry" ><a href="#SUMMARY">Summary</a></td><td id="daEntry" ><a href="#vardefs">Package variables</a></td><td id="daEntry" ><a href="#SYNOPSIS">Synopsis</a></td><td id="daEntry" ><a href="#DESCRIPTION">Description</a></td><td id="daEntry" ><a href="#General">General documentation</a></td></tr></table>
<div id="sep"></div>
<div id="descTitle">Summary</div>
<div id="descArea"><B>MyBio::NGS::JobGraph</B> - Job object, with features</div>
<div id="sep"></div>
<a name="vardefs" id="vardefs"></a><div id="descTitle">Package variables</div>
<div id="descArea"><div id="libEntry">No package variables defined.</div>
</div>
<div id="sep"></div>
<div id="descTitle">Included modules</div>
<div id="descArea"><div id="libEntry"><span id="use">Graph</span> use base qw ( Graph ) </div>
</div>
<div id="sep"></div>
<a name="SYNOPSIS" id="SYNOPSIS"></a>
<div id="descTitle">Synopsis</div>
<div id="descArea"><pre id="podParagraph">    # This is the main transcript object
    # It represents a transcript of a gene
    
    # To initialize 
    my $transcript = <B>MyBio::NGS::JobGraph</B>->new({
        ENSTID         => undef,
        SPECIES        => undef,
        STRAND         => undef,
        CHR            => undef,
        START          => undef,
        STOP           => undef,
        GENE           => undef, # MyBio::Gene
        UTR5           => undef, # MyBio::NGS::JobGraph::UTR5
        CDS            => undef, # MyBio::NGS::JobGraph::CDS
        UTR3           => undef, # MyBio::NGS::JobGraph::UTR3
        CDNA           => undef, # MyBio::NGS::JobGraph::CDNA
        BIOTYPE        => undef,
        INTERNAL_ID    => undef,
        INTERNAL_GID   => undef,
        EXTRA_INFO     => undef,
    });</pre>
</div>
<div id="sep"></div>
<a name="DESCRIPTION" id="DESCRIPTION"></a>
<div id="descTitle">Description</div>
<div id="descArea"><pre id="podParagraph">    The Transcript class describes a transcript of a gene. It has a backreference back to the gene where it belongs
    and contains functional regions such as 5'UTR, CDS and 3'UTR for protein coding genes. If the gene is not protein
    coding then the CDNA region is set and the above remain undefined. It's up to the user to check if the attributes
    are defined or not. The transcript class inherits the SplicedLocus class so it also describes a genomic region
    where Introns and Exons are defined.</pre>
</div>
<div id="sep"></div>
<a name="MethDesc" id="MethDesc"></a>
<div id="mainTitle">Methods description</div>
<div id="sep"></div>
<div id="error">None available.</div>
<a name="MethCode" id="MethCode"></a>
<div id="mainTitle">Methods code</div>
<div id="sep"></div>
<div id="error">No methods available.</div>
<div id="sep"></div>
<a name="General" id="General"></a>
<div id="mainTitle">General documentation</div>
<div id="sep"></div>
<table id="genInfoTitle" width="100%"><tr><td><a name="_pod_EXAMPLES" id="_pod_EXAMPLES"></a>EXAMPLES</td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="genInfoArea"><pre id="podParagraph">    my %transcripts = <B>MyBio::NGS::JobGraph</B>->read_region_info_for_transcripts('FILE',"$database/Ensembl_release_54/CDS_and_3UTR_sequences_CLEAN.txt");</pre></div>
<div id="sep"></div>
<table id="genInfoTitle" width="100%"><tr><td><a name="_pod_AUTHOR - Manolis Maragkakis, Panagiotis Alexiou" id="_pod_AUTHOR - Manolis Maragkakis, Panagiotis Alexiou"></a>AUTHOR - Manolis Maragkakis, Panagiotis Alexiou</td><td align="right"><a href="#TOP">Top</a></td></tr></table>
<div id="genInfoArea">Email <a href="mailto:em.maragkakis@gmail.com" id="podEmail">em.maragkakis@gmail.com</a>, <a href="mailto:pan.alexiou@fleming.gr" id="podEmail">pan.alexiou@fleming.gr</a></div>
<div id="sep"></div>
</body>
</html>
